allen-cell-animated/nucmorph-colorizer
# refactor: Use feature keys instead of names internally

Problem
=======
*Estimated review size: medium, 30-40 minutes. (Most of the changes are variable renames which are easy to validate. Please turn off whitespace changes as well.)*

Closes #181, refactoring the colorizer to use feature keys to separate a feature's identity from its display name.

Solution
========
- Changes Dataset to generate feature keys if none exists (see new `getKeyFromName` method in `data_utils.ts`).
- Updates Datasets to lookup features by key instead of name.
- Updates all other references to features to use keys instead of names.
- The URL will now preferentially use feature keys, but the viewer can also look up by feature name if the key is not found. (This happens on the initial URL load for the scatterplot axes, feature, and thresholds/filters.)

Future work:
- Fix filters tab showing feature keys instead of names? Show a placeholder for unknown keys?

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Open this preview link with feature keys: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-283/#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Fcollection.json&dataset=Mama+Bear&feature=height&filters=volume%3A%25C2%25B5m%25C2%25B3%3A616.112%3A1858.791%2Cheight%3A%25C2%25B5m%3A0.540%3A8.700&range=6.660%2C9.210&scatter-range=all&scatter-x=height&scatter-y=volume
2. Open the same preview link, but with feature names instead of keys to check backwards-compatibility: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-283/#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Fcollection.json&dataset=Mama+Bear&feature=Height&filters=Volume%3A%25C2%25B5m%25C2%25B3%3A616.112%3A1858.791%2CHeight%3A%25C2%25B5m%3A0.540%3A8.700&range=6.660%2C9.210&scatter-range=all&scatter-x=Height&scatter-y=Volume
3. Both should look identical on load.
4. On both, click around the interface and try changing features in various dropdowns.
